### PR TITLE
fix: Copy shoud be able to run under new planner

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

fix: Copy is able to run under new planner

### CODE REVIEW

**Comment by Xuanwo** [comment]
Copy is not running under new planner, this PR will address it.

**Comment by Xuanwo** [comment]
@BohuTANG @sundy-li @andylokandy 

Old planners support the following query:

```sql
COPY INTO mytable
    FROM '@external_stage/path/to/file.csv'
    FILE_FORMAT = (
        type = 'CSV'
        field_delimiter = ','
        record_delimiter = '\n'
        skip_header = 1
    )
    size_limit=10;
```

But our new planner only supports:

```sql
COPY INTO mytable
    FROM @external_stage/path/to/file.csv
    FILE_FORMAT = (
        type = 'CSV'
        field_delimiter = ','
        record_delimiter = '\n'
        skip_header = 1
    )
    size_limit=10;
```

Do we need to support it?

Only `COPY` supports `'@my_stage/abc'`, other SQL like `LIST`, `CREATE`, `DELETE` only support `@my_stage`.

**Comment by Xuanwo** [comment]
Old planner:

```sql
MySQL [(none)]> LIST @xxx;
ERROR 1105 (HY000): Code: 2501, displayText = Unknown stage xxx.
MySQL [(none)]> LIST '@xxx';
ERROR 1105 (HY000): Code: 1005, displayText = sql parser error: Expected @string_literal, found: '@xxx'.

MySQL [(none)]> copy into ontime200 from '@named_external_stage'  PATTERN = 'ontime_200.csv$' FILE_FORMAT = (type = CSV field_delimiter = ','  record_delimiter = '');
ERROR 1105 (HY000): Code: 2501, displayText = Unknown stage named_external_stage.
```

New planner:

```sql
MySQL [(none)]> LIST @xxx;
ERROR 1105 (HY000): Code: 2501, displayText = Unknown stage xxx.
MySQL [(none)]> LIST '@xxx';
ERROR 1105 (HY000): Code: 1005, displayText = sql parser error: Expected @string_literal, found: '@xxx'.

MySQL [(none)]> copy into ontime200 from '@named_external_stage'  PATTERN = 'ontime_200.csv$' FILE_FORMAT = (type = CSV field_delimiter = ','  record_delimiter = '';
ERROR 1105 (HY000): Code: 1005, displayText = error:
  --> SQL:1:51
  |
1 | copy into ontime200 from '@named_external_stage'  PATTERN = 'ontime_200.csv$' FILE_FORMAT = (type = CSV field_delimiter = ','  record_delimiter = ''
  | ----                                              ^^^^^^^ expected `AtString`, <Ident>, `(`, `CONNECTION`, `CREDENTIALS`, or `ENCRYPTION`
  | |
  | while parsing `COPY
                INTO { internalStage | externalStage | externalLocation | [<databas
```

**Comment by BohuTANG** [comment]
Snowflake `COPY` syntax is:
```
COPY INTO [<namespace>.]<table_name>
     FROM { internalStage | externalStage | externalLocation }
[ FILES = ( '<file_name>' [ , '<file_name>' ] [ , ... ] ) ]
```
Where:
```
internalStage ::=
    @[<namespace>.]<int_stage_name>[/<path>]
```

```
externalStage ::=
  @[<namespace>.]<ext_stage_name>[/<path>]
```

```
externalLocation (for Amazon S3) ::=
  's3://<bucket>[/<path>]' -- here is the point
...
```
https://docs.snowflake.com/en/sql-reference/sql/copy-into-table.html#examples

It's very clear what we would support.

The `LIST` syntax is the same too:
https://docs.snowflake.com/en/sql-reference/sql/list.html

**Comment by Xuanwo** [comment]
> It's very clear what we would support.

We does support `copy into ontime200 from 's3://bucket/path/to/file'`

The problem raised in this PR does we need to support `copy into ontime200 from '@my_stage'`.

Snowflake seems don't support it.

**Comment by BohuTANG** [comment]
> copy into ontime200 from '@my_stage'

It should be `copy into ontime200 from @my_stage`

**Comment by Xuanwo** [comment]
> It should be `copy into ontime200 from @my_stage`

Got it. I will update the docs to make sure the behavior is consistent as expected.

**Comment by BohuTANG** [comment]
As snowflake note:
```
If the stage name or path includes spaces or special characters, it must be enclosed in single quotes (e.g. '@"my stage"' for a stage named "my stage").
```

**Comment by Xuanwo** [comment]
> As snowflake note:
> 
> ```
> If the stage name or path includes spaces or special characters, it must be enclosed in single quotes (e.g. '@"my stage"' for a stage named "my stage").
> ```
> 
> https://docs.snowflake.com/en/sql-reference/sql/list.html

I will implement this feature in a separate issue.

File: query/src/sql/planner/binder/copy.rs [hunk 1]
```diff
@@ -445,6 +452,8 @@ impl<'a> Binder {
         let query = self
             .bind_statement(bind_context, &Statement::Query(Box::new(src_query.clone())))
             .await?;
+        // Make sure query has been optimized.
+        let query = optimize(self.ctx.clone(), query)?;
```
Comment (leiysky) [inline query/src/sql/planner/binder/copy.rs:456]: https://github.com/datafuselabs/databend/blob/c774f800d5e458062c5204d7822668125ff66ba9/query/src/sql/optimizer/mod.rs#L66

IMO it's better to enable optimization for `CopyPlan` here.

Comment (Xuanwo) [reply query/src/sql/planner/binder/copy.rs]: Addressed.

**Review by BohuTANG — APPROVED** [review]
LGTM

**Comment by BohuTANG** [comment]
@soyeric128 

Please update the documentation about the `stage` this PR merged:

Old:
```
copy into ontime200 from '@named_external_stage' ...
```

New:
```
copy into ontime200 from @named_external_stage ...
```

**Comment by soyeric128** [comment]
> @soyeric128
> 
> Please update the documentation about the `stage` after this PR merged:
> 
> Old:
> 
> ```
> copy into ontime200 from '@named_external_stage' ...
> ```
> 
> New:
> 
> ```
> copy into ontime200 from @named_external_stage ...
> ```

Does it also apply to the named internal stages?

**Pull request merged**